runtime/src/iree/schemas/bytecode_module_def.fbs

```diff
@@ -94,13 +94,20 @@ union CompressionTypeDef {
 }
 
 // Read-only data segment.
+// The data may be embedded directly in the flatbuffer or point to a reference
+// relative to the flatbuffer in memory.
```

Comment:
High level question before I dig into the details: what about sharding across multiple files / non-contiguous memory locations? Is that something we should support? If so, will that be compatible with this offset-in-memory segmentation?

I can imagine cases, particularly on mobile and the web, where we'd want to fetch and warmup executable code and entry points quickly, while allowing parts of the program to be fetched asynchronously (or even lazily).

Reply:
oh and if you care about performance on web you want to statically link your executable code such that it gets JIT'ed with the rest of the runtime - you can still use vmfbs for all the rest of the stuff (that'd be what you fetch in parallel) or just go the whole way with emitc/llvm to avoid vmfbs entirely; then if you have big rodata you'd do as above with `@set_big_params(buffer)`. You wouldn't want to ship the executable code in a vmfb as that puts JIT'ing the runtime, fetching the vmfb with the wasm executables (even if you exclude constants), and running the vm initializer in the critical path before the browser even sees the wasm executable blobs. That said, warm start in a browser with caching will be fine for users who aren't worried about cold start (things not on homepages of top 10% web properties).

Reply:
drive-by:  Can we have the .vmfb be a vmfb/  directory?  We are considering cases where the vmfb file is nearing multi-terabyte sizes and is likely hosted on a shared filesystem like lustre. Each node will fetch parts of the "vmfb" based on its rank without having to do an "unzip" on a massive file.

Reply:
The zip metadata is just for debugging - we don't actually use it at runtime. The vmfbs are just like disk images and can be treated as such when on a random-access block-based network filesystem. A pass we still need to write would ensure constants are clustered in the file based on usage to make that more efficient - today they're mostly in the order from the original model and work out that way but it's usually just by chance.

I'm not sure what lustre does, but if it's doing file/inode-based replication instead of block-based it would be like the cases above where the user or some layer in the compiler would have to spit out multiple vmfbs or a single vmfb for all shards + multiple raw binary files. It's up to the user/layers above to do the transfer/loading/etc, though. We could grow a layer above the core IREE compiler/runtime that does this kind of stuff in the future but don't have anything there today.

If using a technique like jax pmap where there's explicit user placement this may just fall out naturally if each shard was compiled to independent vmfbs as each shard would get its own relevant data. It'd be really nice to do this automatically using something in the stream dialect after we perform placement and we could have an option to produce multiple vmfbs - it's just an extension of the clustering optimization that'd be useful on local devices.

Lots of design space there to explore but probably something we'll have to look at after we get multiple devices on a single node working :)

Reply:
Gotcha. For more framework-level features that are left up to the user, I wonder what we could contribute without taking on too much complexity ourselves. Just brainstorming:

* A doc (https://google.github.io/iree/extensions/?) listing what the core project offers and what higher layers can add
* Sample projects (either authored by us or from the open source community) demonstrating things like splitting a program into pieces
* A framework built on IREE, like https://github.com/tensorflow/addons (could have one for embedded, one for servers, etc.)... but that gets really messy

Reply:
yeah, it'd be awesome to have some basic stuff in optional modules - I think the biggest design point issue is that most of these reach up into the frameworks the models are coming from. I'm not sure there's much extra needed functionality-wise besides documentation outside of datacenter stuff because datacenter stuff gets weird and specialized fast. I'm not sure if there's much we could do besides have the right APIs and let the integrations handle things. More users trying to do that stuff will help us figure out what we can actually provide without just writing one-off stuff that isn't actually usable. Example code using our APIs would be good, though, and is a much lower bar than trying to productionize something for uses we don't understand or want to focus on.